src/oge/helpers.py

```diff
@@ -0,0 +1,358 @@
+            ba_name_to_ba_code["ba_code"],
+        )
+    )
+
+    # specify a ba code for certain utilities
+    utility_as_ba_code = pd.read_csv(
+        reference_table_folder("utility_name_ba_code_map.csv")
+    )
+    utility_as_ba_code = dict(
+        zip(
+            utility_as_ba_code["name"],
+            utility_as_ba_code["ba_code"],
+        )
+    )
+
+    # fill missing BA codes first based on the BA name, then utility name, then on the transmisison owner name
+    plant_ba["balancing_authority_code_eia"] = plant_ba[
+        "balancing_authority_code_eia"
+    ].fillna(plant_ba["balancing_authority_name_eia"].map(ba_name_to_ba_code))
+    plant_ba["balancing_authority_code_eia"] = plant_ba[
+        "balancing_authority_code_eia"
+    ].fillna(plant_ba["balancing_authority_name_eia"].map(utility_as_ba_code))
+    plant_ba["balancing_authority_code_eia"] = plant_ba[
+        "balancing_authority_code_eia"
+    ].fillna(plant_ba["utility_name_eia"].map(utility_as_ba_code))
+    plant_ba["balancing_authority_code_eia"] = plant_ba[
+        "balancing_authority_code_eia"
+    ].fillna(plant_ba["transmission_distribution_owner_name"].map(utility_as_ba_code))
+
+    # rename the ba column
+    plant_ba = plant_ba.rename(columns={"balancing_authority_code_eia": "ba_code"})
+
+    plant_ba["ba_code"] = plant_ba["ba_code"].replace("None", np.NaN)
+
+    # get a list of all of the BAs that retired prior to the current year
+    retired_bas = load_data.load_ba_reference()[["ba_code", "retirement_date"]]
+    retired_bas = list(
+        retired_bas.loc[
+            retired_bas["retirement_date"].dt.year < year, "ba_code"
+        ].unique()
+    )
+    # if there are any plants that have been assigned to a retired BA, set its BA code as missing
+    plant_ba.loc[plant_ba["ba_code"].isin(retired_bas), "ba_code"] = np.NaN
+
+    # for plants without a BA code assign the miscellaneous BA code based on the state
+    plant_ba["ba_code"] = plant_ba["ba_code"].fillna(plant_ba["state"] + "MS")
+
+    # add a physical ba code based on the owner of the transmission system
+    plant_ba["ba_code_physical"] = plant_ba["ba_code"]
+    plant_ba.update(
+        pd.DataFrame(
+            {
+                "ba_code_physical": plant_ba["transmission_distribution_owner_name"]
+                .map(utility_as_ba_code)
+                .to_list()
+            }
+        )
+    )
+
+    # update based on mapping table when ambiguous
+    physical_ba = pd.read_csv(
+        reference_table_folder("physical_ba.csv"), dtype=get_dtypes()
+    )
+    plant_ba = plant_ba.merge(
+        physical_ba,
+        how="left",
+        on=["ba_code", "transmission_distribution_owner_name"],
+        suffixes=("", "_map"),
+        validate="m:1",
+    )
+    plant_ba.update(
+        pd.DataFrame({"ba_code_physical": plant_ba["ba_code_physical_map"].to_list()})
+    )
+
+    return plant_ba
+
+
+def map_eia_plant_id_to_ba_code():
```

Comment:
It seems like this will work, but this solution (including the new `utility` module) seems to add a lot of unneeded complexity just to  avoid running `create_plant_ba_table()` (which only takes 1 to 1.5 seconds to load a pretty small dataframe) and merging it into the dataframe we're validating a handful of times. 

It seems like it would have been far simpler to just run
```
df = df.merge(create_plant_ba_table(year)[["plant_id_eia,"ba_code"]], how="left", on="plant_id_eia", validate="m:1")
```
and pass `year` to wherever we needed it (or access it though a global variable.

Reply:
Sure. I will do what you propose